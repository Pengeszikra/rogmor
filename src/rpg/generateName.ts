import { rnd } from "./rpg";

export const ROGMOR_NAME_SOURCE = 'proedmorainowerialaenderomanulnivosudanterill';

export const generateName = (sour = ROGMOR_NAME_SOURCE) => {
  const name:string = Array.from(
      {length:2 + rnd(3)}, 
      (s:number = rnd(sour.length)) => sour.slice(s, s + 2 + rnd(3))
    )
    .join('')
  ;
  const [capital, ...rest] = name.split('');
  return [capital.toUpperCase(),...rest].join('');
}
