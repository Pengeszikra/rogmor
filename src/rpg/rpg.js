export const leveling = (level, mod = 0.5, pow = 4, madd = 2) => Math.round(
  mod * Math.pow(level, pow) + (madd * level)
);

export const rnd = dice => dice * Math.random() | 0;

export const dice = roll => rnd(roll - 1) + 1;

// export const improved = ( attribute = 1, percent = 0.2 ) => (attribute + (Math.random() * percent * attribute)) | 0;
export const improved = ( attribute = 1 ) => attribute * 1.15 | 0;

export const shuffle = _ => Math.random() > 0.5 ? 1 : -1;

export const pickOne = arr => arr[rnd(arr.length)];

export const uid =  _ => Math.random().toString(32).slice(-8);

export const amount = n => Array(n).fill().map((_, index) => index);