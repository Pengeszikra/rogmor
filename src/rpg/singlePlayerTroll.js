import {actionFactory, kebabToCamelCase} from 'react-troll';
import { physicalStrike, socialTalk, soulSkill } from '../gui/battleSaga';

export const [getActionsLookup, action] = actionFactory(kebabToCamelCase);

export const initialState = {
 round : 0,
 hero: null,
 game: {},
 entities: {},
 focus: null,
};

export const 
  SET_HERO = action('set-hero'),
  MOD_HERO = action('mod-hero'),
  NEXT_ROUND = action('next-round'),
  SET_GAME_STATE = action('set-game-state'),
  SETUP_ENTITIES = action('setup-entities'),
  MOD_ENTITI = action('mod-entiti'),
  ADD_ENTITI = action('add-entiti'),
  REMOVE_ENTITI = action('remove-entiti'),
  FOCUS_ON = action('focus-on'),
  FIGHT = action('fight'),
  SKILL = action('skill'),
  TALK  = action('talk')
;

export const gameReducer = (state, {type, payload}) => {
  switch (type) {
    case SET_HERO: return {...state, hero: payload};
    case MOD_HERO: return {...state, hero: state.hero |> payload};
    case NEXT_ROUND: return {...state, round:  state.round |> payload};
    case SET_GAME_STATE: return {...state, game: state.game |> payload};
    case SETUP_ENTITIES: return {...state, entities: payload};
    case MOD_ENTITI: {
      const {entities} = state;
      return entities[payload?.uid] 
        ? {...state, entities: {...entities, [uid]: payload} }
        : state;
    };
    case FOCUS_ON: return {...state, focus: payload};
    case FIGHT: return state |> fightRound |> useFullCheck;
    case SKILL: return state |> skillRound |> useFullCheck;
    case TALK:  return state |> talkRound  |> useFullCheck;
    default: return state;
  }
};

const lostFocus = s => s

const useFullCheck = ({focus, entities, ...rest}) => {
  const {staminaState, willpowerState, merryState} = entities[focus];
  return staminaState > 0 
      && willpowerState > 0 
      && merryState > 0
      ? {focus, entities, ...rest} : {focus: null, entities, ...rest}; 
}

const interactionRound = interaction => ({hero, entities, round, focus, ...rest}) => {
  const [hMod, npcMod] = interaction(hero, entities[focus], round);
  return {...rest, hero:hMod, round: round + 1, focus, entities:{...entities, [npcMod.uid]:npcMod}};
};

const fightRound = physicalStrike |> interactionRound;
const skillRound = soulSkill |> interactionRound;
const  talkRound = socialTalk |> interactionRound;

