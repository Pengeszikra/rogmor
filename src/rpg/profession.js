import {rnd, leveling} from './rpg';

export const profTypes = {
    fighter:      { profession:'Fighter',      physique: 7, reaction: 4, soul: 2, liaison: 4, stamina: 55, willpower: 45, merry: 40 },
    priest:       { profession:'Priest',       physique: 3, reaction: 3, soul: 7, liaison: 8, stamina: 35, willpower: 65, merry: 40 },
    paladin:      { profession:'Paladin',      physique: 6, reaction: 3, soul: 6, liaison: 6, stamina: 50, willpower: 50, merry: 40 },
    knight:       { profession:'Knight',       physique: 7, reaction: 3, soul: 3, liaison: 5, stamina: 60, willpower: 50, merry: 40 },
    assasin:      { profession:'Assasin',      physique: 5, reaction: 7, soul: 1, liaison: 3, stamina: 45, willpower: 55, merry: 40 },
    rogue:        { profession:'Rogue',        physique: 4, reaction: 6, soul: 3, liaison: 5, stamina: 55, willpower: 40, merry: 40 },
    mage:         { profession:'Mage',         physique: 3, reaction: 4, soul: 6, liaison: 2, stamina: 25, willpower: 70, merry: 40 },
    captain:      { profession:'Captain',      physique: 5, reaction: 5, soul: 5, liaison: 6, stamina: 55, willpower: 55, merry: 40 },
    noble:        { profession:'Noble',        physique: 4, reaction: 5, soul: 6, liaison: 7, stamina: 45, willpower: 60, merry: 40 },
    headHunter:   { profession:'Head hunter',  physique: 5, reaction: 7, soul: 3, liaison: 4, stamina: 60, willpower: 45, merry: 40 },
    witch:        { profession:'Witch',        physique: 2, reaction: 5, soul: 6, liaison: 2, stamina: 25, willpower: 65, merry: 40 },
    valkur:       { profession:'Valkur',       physique: 5, reaction: 6, soul: 4, liaison: 2, stamina: 60, willpower: 60, merry: 40 },
    guard:        { profession:'Guard',        physique: 6, reaction: 5, soul: 2, liaison: 4, stamina: 65, willpower: 40, merry: 40 },
    gladiator:    { profession:'Gladiator',    physique: 7, reaction: 5, soul: 2, liaison: 2, stamina: 75, willpower: 25, merry: 40 },
    barbarian:    { profession:'Barbarian',    physique: 8, reaction: 6, soul: 2, liaison: 2, stamina: 80, willpower: 35, merry: 40 },
    wizard:       { profession:'Wizard',       physique: 1, reaction: 2, soul: 8, liaison: 2, stamina: 25, willpower: 80, merry: 40 },
    sage:         { profession:'Sage',         physique: 4, reaction: 5, soul: 6, liaison: 5, stamina: 45, willpower: 60, merry: 40 },
    samurai:      { profession:'Samurai',      physique: 6, reaction: 7, soul: 4, liaison: 5, stamina: 70, willpower: 55, merry: 40 },
    ninja:        { profession:'Ninja',        physique: 4, reaction: 8, soul: 3, liaison: 3, stamina: 55, willpower: 45, merry: 40 },
    warlord:      { profession:'Warlord',      physique: 4, reaction: 5, soul: 6, liaison: 8, stamina: 45, willpower: 60, merry: 40 },
    bishop:       { profession:'Bishop',       physique: 4, reaction: 5, soul: 6, liaison: 8, stamina: 45, willpower: 60, merry: 40 },
    tracker:      { profession:'Tracker',      physique: 4, reaction: 5, soul: 6, liaison: 5, stamina: 45, willpower: 60, merry: 40 },
    dementor:     { profession:'Dementor',     physique: 6, reaction: 4, soul: 1, liaison: 1, stamina: 75, willpower: 70, merry: 40 },
    necromanta:   { profession:'Necromanta',   physique: 2, reaction: 5, soul: 6, liaison: 1, stamina: 45, willpower: 60, merry: 40 },
    summoner:     { profession:'Summoner',     physique: 3, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    shaman:       { profession:'Shaman',       physique: 4, reaction: 5, soul: 6, liaison: 6, stamina: 45, willpower: 60, merry: 40 },
    thief:        { profession:'Thief',        physique: 4, reaction: 5, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    rouge:        { profession:'Rouge',        physique: 4, reaction: 5, soul: 6, liaison: 5, stamina: 45, willpower: 60, merry: 40 },
    cavalier:     { profession:'Cavalier',     physique: 6, reaction: 5, soul: 6, liaison: 4, stamina: 45, willpower: 60, merry: 40 },
    pirate:       { profession:'Pirate',       physique: 4, reaction: 5, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    inquisitor:   { profession:'Inquisitor',   physique: 4, reaction: 5, soul: 6, liaison: 4, stamina: 45, willpower: 60, merry: 40 },
    icelander:    { profession:'Icelander',    physique: 7, reaction: 6, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    heretic:      { profession:'Heretic',      physique: 4, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    nomad:        { profession:'Nomad',        physique: 4, reaction: 5, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    bard:         { profession:'Bard',         physique: 4, reaction: 5, soul: 6, liaison: 6, stamina: 45, willpower: 60, merry: 40 },
    warlock:      { profession:'Warlock',      physique: 4, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    monk:         { profession:'Monk',         physique: 6, reaction: 6, soul: 6, liaison: 5, stamina: 45, willpower: 60, merry: 40 },
    druid:        { profession:'Druid',        physique: 4, reaction: 5, soul: 6, liaison: 4, stamina: 45, willpower: 60, merry: 40 },
    illusionist:  { profession:'Illusionist',  physique: 3, reaction: 6, soul: 5, liaison: 5, stamina: 35, willpower: 50, merry: 40 },
    wardancer:    { profession:'Wardancer',    physique: 5, reaction: 7, soul: 6, liaison: 2, stamina: 60, willpower: 40, merry: 40 },
    runeSmith:    { profession:'Rune smith',   physique: 4, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    merchant:     { profession:'Merchant',     physique: 3, reaction: 3, soul: 6, liaison: 6, stamina: 55, willpower: 65, merry: 40 },
    explorer:     { profession:'Explorer',     physique: 4, reaction: 5, soul: 6, liaison: 7, stamina: 45, willpower: 60, merry: 40 },
    templar:      { profession:'Templar',      physique: 6, reaction: 3, soul: 6, liaison: 5, stamina: 60, willpower: 55, merry: 40 },
    spy:          { profession:'Spy',          physique: 4, reaction: 5, soul: 6, liaison: 5, stamina: 45, willpower: 60, merry: 40 },
    nun:          { profession:'Nun',          physique: 4, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    mercenary:    { profession:'Mercenary',    physique: 4, reaction: 5, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    duelist:      { profession:'Duelist',      physique: 3, reaction: 8, soul: 5, liaison: 5, stamina: 50, willpower: 60, merry: 40 },
    beastmaster:  { profession:'Beastmaster',  physique: 6, reaction: 5, soul: 7, liaison: 1, stamina: 45, willpower: 70, merry: 40 },
    sorcerer:     { profession:'Sorcerer',     physique: 3, reaction: 5, soul: 6, liaison: 2, stamina: 45, willpower: 60, merry: 40 },
    archer:       { profession:'Archer',       physique: 4, reaction: 7, soul: 6, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    psionic:      { profession:'Psionic',      physique: 4, reaction: 4, soul: 7, liaison: 3, stamina: 45, willpower: 60, merry: 40 },
    alchemist:    { profession:'Alchemist',    physique: 3, reaction: 5, soul: 7, liaison: 2, stamina: 35, willpower: 70, merry: 40 },
    enginer:      { profession:'Enginer',      physique: 3, reaction: 5, soul: 4, liaison: 4, stamina: 35, willpower: 65, merry: 40 },
    antiPaladin:  { profession:'Anti Paladin', physique: 7, reaction: 3, soul: 1, liaison: 5, stamina: 60, willpower: 60, merry: 10 },
    chaosKnigh:   { profession:'Chaos Knigh',  physique: 7, reaction: 4, soul: 6, liaison: 1, stamina: 50, willpower: 60, merry: 10 },
    chief:        { profession:'Chief',        physique: 5, reaction: 5, soul: 4, liaison: 7, stamina: 55, willpower: 60, merry: 40 },
    maruder:      { profession:'Maruder',      physique: 7, reaction: 4, soul: 2, liaison: 4, stamina: 60, willpower: 40, merry: 40 },
    rich:         { profession:'Rich',         physique: 3, reaction: 4, soul: 3, liaison: 8, stamina: 30, willpower: 70, merry: 30 },
    shapechanger: { profession:'Shapechanger', physique: 6, reaction: 4, soul: 4, liaison: 2, stamina: 55, willpower: 45, merry: 30 },

    // mentor, diviner, detective, seer
}

// https://en.wikipedia.org/wiki/Character_class_(Dungeons_%26_Dragons)
// https://wfrp1e.fandom.com/wiki/List_Of_Advanced_Careers

/// const professions = Object.keys(profTypes).map( key => profTypes[key]);
const professions = Object.entries(profTypes).map( ([_, prof]) => prof);

export default (level = 1, profession = professions[rnd(professions.length)] ) => {
  const leveled = {
    physique   : leveling(level, profession.physique  / 10, 2),
    reaction   : leveling(level, profession.reaction  / 10, 2),
    soul       : leveling(level, profession.soul      / 10, 2),
    liaison    : leveling(level, profession.liaison   / 10, 2),
    stamina    : leveling(level, profession.stamina   / 10, 2),
    willpower  : leveling(level, profession.willpower / 10, 2),
    merry      : leveling(level, profession.merry     / 10, 2),
    profession : profession.profession,
  };
  leveled.staminaState = leveled.stamina;
  leveled.willpowerState = leveled.willpower;
  leveled.merryState = leveled.merry;
  return ({...leveled, level, type: profession});
}