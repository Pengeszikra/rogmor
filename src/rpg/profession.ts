import { rnd, leveling, uid, pickOne } from './rpg';

export interface Ability { 
  title:string;
  body: number;
  reaction: number;
  soul: number;
  popular: number;
  stamina: number;
  will: number;
  joyful: number; 
}

export interface Condition {
  staminaState: number;
  willState: number;
  joyfulState: number;
}

export enum Team { PLAYER, GOOD, BAD, UGLY };

export interface MobTraits {  
  level: number;
  professionType: ProfessionKey;
  ability: Ability;
  condition: Condition;
}
export interface MobPerson {  
  name: string;
  avatar: number;
  coord: number;
  uid: string;
  team: Team;
}

export interface Mob extends MobTraits, MobPerson {}

export type ProfessionKey =
  'fighter' |'priest' |'paladin' |'knight' |'assasin' |'rogue' |'mage' |'captain' |'noble' |
  'headHunter' |'witch' |'valkur' |'guard' |'gladiator' |'barbarian' |'wizard' |'sage' |
  'samurai' |'ninja' |'warlord' |'bishop' |'tracker' |'dementor' |'necromanta' |'summoner' |
  'shaman' |'thief' |'rouge' |'cavalier' |'pirate' |'inquisitor' |'icelander' |'heretic' |
  'nomad' |'bard' |'warlock' |'monk' |'druid' |'illusionist' |'wardancer' |'runeSmith' |
  'merchant' |'explorer' |'templar' |'spy' |'nun' |'mercenary' |'duelist' |'beastmaster' |
  'sorcerer' |'archer' |'psionic' |'alchemist' |'enginer' |'antiPaladin' |'chaosKnigh' |
  'chief' |'maruder' |'rich' |'shapechanger' 

export type ProfessionLookup = Record<ProfessionKey, Ability>; 

export const professionLookup:ProfessionLookup = {
    fighter:      { title:'Fighter',      body: 7, reaction: 4, soul: 2, popular: 4, stamina: 55, will: 45, joyful: 40 },
    priest:       { title:'Priest',       body: 3, reaction: 3, soul: 7, popular: 8, stamina: 35, will: 65, joyful: 40 },
    paladin:      { title:'Paladin',      body: 6, reaction: 3, soul: 6, popular: 6, stamina: 50, will: 50, joyful: 40 },
    knight:       { title:'Knight',       body: 7, reaction: 3, soul: 3, popular: 5, stamina: 60, will: 50, joyful: 40 },
    assasin:      { title:'Assasin',      body: 5, reaction: 7, soul: 1, popular: 3, stamina: 45, will: 55, joyful: 40 },
    rogue:        { title:'Rogue',        body: 4, reaction: 6, soul: 3, popular: 5, stamina: 55, will: 40, joyful: 40 },
    mage:         { title:'Mage',         body: 3, reaction: 4, soul: 6, popular: 2, stamina: 25, will: 70, joyful: 40 },
    captain:      { title:'Captain',      body: 5, reaction: 5, soul: 5, popular: 6, stamina: 55, will: 55, joyful: 40 },
    noble:        { title:'Noble',        body: 4, reaction: 5, soul: 6, popular: 7, stamina: 45, will: 60, joyful: 40 },
    headHunter:   { title:'Head hunter',  body: 5, reaction: 7, soul: 3, popular: 4, stamina: 60, will: 45, joyful: 40 },
    witch:        { title:'Witch',        body: 2, reaction: 5, soul: 6, popular: 2, stamina: 25, will: 65, joyful: 40 },
    valkur:       { title:'Valkur',       body: 5, reaction: 6, soul: 4, popular: 2, stamina: 60, will: 60, joyful: 40 },
    guard:        { title:'Guard',        body: 6, reaction: 5, soul: 2, popular: 4, stamina: 65, will: 40, joyful: 40 },
    gladiator:    { title:'Gladiator',    body: 7, reaction: 5, soul: 2, popular: 2, stamina: 75, will: 25, joyful: 40 },
    barbarian:    { title:'Barbarian',    body: 8, reaction: 6, soul: 2, popular: 2, stamina: 80, will: 35, joyful: 40 },
    wizard:       { title:'Wizard',       body: 1, reaction: 2, soul: 8, popular: 2, stamina: 25, will: 80, joyful: 40 },
    sage:         { title:'Sage',         body: 4, reaction: 5, soul: 6, popular: 5, stamina: 45, will: 60, joyful: 40 },
    samurai:      { title:'Samurai',      body: 6, reaction: 7, soul: 4, popular: 5, stamina: 70, will: 55, joyful: 40 },
    ninja:        { title:'Ninja',        body: 4, reaction: 8, soul: 3, popular: 3, stamina: 55, will: 45, joyful: 40 },
    warlord:      { title:'Warlord',      body: 4, reaction: 5, soul: 6, popular: 8, stamina: 45, will: 60, joyful: 40 },
    bishop:       { title:'Bishop',       body: 4, reaction: 5, soul: 6, popular: 8, stamina: 45, will: 60, joyful: 40 },
    tracker:      { title:'Tracker',      body: 4, reaction: 5, soul: 6, popular: 5, stamina: 45, will: 60, joyful: 40 },
    dementor:     { title:'Dementor',     body: 6, reaction: 4, soul: 1, popular: 1, stamina: 75, will: 70, joyful: 40 },
    necromanta:   { title:'Necromanta',   body: 2, reaction: 5, soul: 6, popular: 1, stamina: 45, will: 60, joyful: 40 },
    summoner:     { title:'Summoner',     body: 3, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    shaman:       { title:'Shaman',       body: 4, reaction: 5, soul: 6, popular: 6, stamina: 45, will: 60, joyful: 40 },
    thief:        { title:'Thief',        body: 4, reaction: 5, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    rouge:        { title:'Rouge',        body: 4, reaction: 5, soul: 6, popular: 5, stamina: 45, will: 60, joyful: 40 },
    cavalier:     { title:'Cavalier',     body: 6, reaction: 5, soul: 6, popular: 4, stamina: 45, will: 60, joyful: 40 },
    pirate:       { title:'Pirate',       body: 4, reaction: 5, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    inquisitor:   { title:'Inquisitor',   body: 4, reaction: 5, soul: 6, popular: 4, stamina: 45, will: 60, joyful: 40 },
    icelander:    { title:'Icelander',    body: 7, reaction: 6, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    heretic:      { title:'Heretic',      body: 4, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    nomad:        { title:'Nomad',        body: 4, reaction: 5, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    bard:         { title:'Bard',         body: 4, reaction: 5, soul: 6, popular: 6, stamina: 45, will: 60, joyful: 40 },
    warlock:      { title:'Warlock',      body: 4, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    monk:         { title:'Monk',         body: 6, reaction: 6, soul: 6, popular: 5, stamina: 45, will: 60, joyful: 40 },
    druid:        { title:'Druid',        body: 4, reaction: 5, soul: 6, popular: 4, stamina: 45, will: 60, joyful: 40 },
    illusionist:  { title:'Illusionist',  body: 3, reaction: 6, soul: 5, popular: 5, stamina: 35, will: 50, joyful: 40 },
    wardancer:    { title:'Wardancer',    body: 5, reaction: 7, soul: 6, popular: 2, stamina: 60, will: 40, joyful: 40 },
    runeSmith:    { title:'Rune smith',   body: 4, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    merchant:     { title:'Merchant',     body: 3, reaction: 3, soul: 6, popular: 6, stamina: 55, will: 65, joyful: 40 },
    explorer:     { title:'Explorer',     body: 4, reaction: 5, soul: 6, popular: 7, stamina: 45, will: 60, joyful: 40 },
    templar:      { title:'Templar',      body: 6, reaction: 3, soul: 6, popular: 5, stamina: 60, will: 55, joyful: 40 },
    spy:          { title:'Spy',          body: 4, reaction: 5, soul: 6, popular: 5, stamina: 45, will: 60, joyful: 40 },
    nun:          { title:'Nun',          body: 4, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    mercenary:    { title:'Mercenary',    body: 4, reaction: 5, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    duelist:      { title:'Duelist',      body: 3, reaction: 8, soul: 5, popular: 5, stamina: 50, will: 60, joyful: 40 },
    beastmaster:  { title:'Beastmaster',  body: 6, reaction: 5, soul: 7, popular: 1, stamina: 45, will: 70, joyful: 40 },
    sorcerer:     { title:'Sorcerer',     body: 3, reaction: 5, soul: 6, popular: 2, stamina: 45, will: 60, joyful: 40 },
    archer:       { title:'Archer',       body: 4, reaction: 7, soul: 6, popular: 3, stamina: 45, will: 60, joyful: 40 },
    psionic:      { title:'Psionic',      body: 4, reaction: 4, soul: 7, popular: 3, stamina: 45, will: 60, joyful: 40 },
    alchemist:    { title:'Alchemist',    body: 3, reaction: 5, soul: 7, popular: 2, stamina: 35, will: 70, joyful: 40 },
    enginer:      { title:'Enginer',      body: 3, reaction: 5, soul: 4, popular: 4, stamina: 35, will: 65, joyful: 40 },
    antiPaladin:  { title:'Anti Paladin', body: 7, reaction: 3, soul: 1, popular: 5, stamina: 60, will: 60, joyful: 10 },
    chaosKnigh:   { title:'Chaos Knigh',  body: 7, reaction: 4, soul: 6, popular: 1, stamina: 50, will: 60, joyful: 10 },
    chief:        { title:'Chief',        body: 5, reaction: 5, soul: 4, popular: 7, stamina: 55, will: 60, joyful: 40 },
    maruder:      { title:'Maruder',      body: 7, reaction: 4, soul: 2, popular: 4, stamina: 60, will: 40, joyful: 40 },
    rich:         { title:'Rich',         body: 3, reaction: 4, soul: 3, popular: 8, stamina: 30, will: 70, joyful: 30 },
    shapechanger: { title:'Shapechanger', body: 6, reaction: 4, soul: 4, popular: 2, stamina: 55, will: 45, joyful: 30 },
}

// https://en.wikipedia.org/wiki/Character_class_(Dungeons_%26_Dragons)
// https://wfrp1e.fandom.com/wiki/List_Of_Advanced_Careers

export const professionList = Object.values(professionLookup);

export const traitsFactory = (level:number = 1,  professionType:ProfessionKey) => {

  const ability:Ability = professionLookup?.[professionType] ?? pickOne(professionList);

  const stamina = leveling(level, ability.stamina  / 10, 2);
  const will    = leveling(level, ability.will     / 10, 2);
  const joyful  = leveling(level, ability.joyful   / 10, 2);

  const mobTraits:MobTraits = {
    level,
    professionType,
    ability: {
      title        : ability.title,
      body         : leveling(level, ability.body     / 10, 2),
      reaction     : leveling(level, ability.reaction / 10, 2),
      soul         : leveling(level, ability.soul     / 10, 2),
      popular      : leveling(level, ability.popular  / 10, 2),
      stamina,
      will,
      joyful,
    },
    condition: {
      staminaState : stamina,
      willState    : will,
      joyfulState  : joyful,
    }
  };

  return mobTraits;
}

export const mobFactory = (name, avatar, coord, uid, team, traits:MobTraits):Mob => ({
  name,
  avatar,
  coord,
  uid,
  team,
  ...traits
});

export const increaseLevel = (amount:number = 1) => ({
  level, 
  professionType, 
  ability, 
  condition, 
  ...rest
}:Mob):Mob => ({
  ...traitsFactory(level + amount, professionType),
  ...rest
});

export const oneLevelUp = increaseLevel(1);