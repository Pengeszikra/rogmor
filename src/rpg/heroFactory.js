import generateName from "./generateName";
import profession from "./profession";

export const heroFactory = (heroId, lvl = 1) => ({heroId, name: generateName(), ...profession(lvl)});
