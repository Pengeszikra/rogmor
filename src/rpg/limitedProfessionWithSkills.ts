import { ProfessionKey } from "./profession";

export const limitedProfessionWithSkills:ProfessionKey[] = [
  'chaosKnight',
  'bishop',
  'icelander',
  'ninja',
  'samurai',
  'merchant',
];