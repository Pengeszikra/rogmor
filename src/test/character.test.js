import profession, { profTypes } from "../rpg/profession";
import { useEntitiReducer } from "../rpg/useEntitiReduce";

import { shallow, configure } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
configure({ adapter: new Adapter() });

test ('create fighter', () => {  
  expect (
    profTypes.fighter
  ).toStrictEqual(
    { profession:'Fighter',      physique: 7, reaction: 4, soul: 2, liaison: 4, stamina: 55, willpower: 45, merry: 40 }
  );
});


test ('create lvl 1 fighter', () => {
  expect (
    profession(1, profTypes.fighter, _ => 42)
  ).toStrictEqual(
    {
        level: 1,
        liaison: 2,
        merry: 6,
        merryState: 6,
        physique: 3,
        profession: "Fighter",
        reaction: 2,
        soul: 2,
        stamina: 8,
        staminaState: 8,
        type: {
          liaison: 4,
          merry: 40,
          physique: 7,
          profession: "Fighter",
          reaction: 4,
          soul: 2,
          stamina: 55,
          willpower: 45,
        },
        uid: 42,
        willpower: 7,
        willpowerState: 7,
      }
  );
});

test ('useEntitiReduce', () => {
  const Unit = () => {
    const [unit, actions] = useEntitiReducer()
    return <pre>{JSON.stringify(unit, null, 2)}</pre>
  }
  
  expect(shallow(
    <Unit />
  )).toMatchSnapshot();

});