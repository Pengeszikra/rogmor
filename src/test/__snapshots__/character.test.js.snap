// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useEntitiReduce 1`] = `
<pre>
  {
  "physique": 2,
  "reaction": 2,
  "soul": 2,
  "liaison": 2,
  "stamina": 3,
  "willpower": 7,
  "merry": 7,
  "profession": "Children",
  "uid": "_entitiy_",
  "staminaState": 3,
  "willpowerState": 7,
  "merryState": 7,
  "level": 1,
  "type": {
    "profession": "Children",
    "physique": 1,
    "reaction": 1,
    "soul": 2,
    "liaison": 1,
    "stamina": 5,
    "willpower": 50,
    "merry": 50
  }
}
</pre>
`;
